File: cpp/155.cpp
#include<bits/stdc++.h>
using std::vector;

class MinStack {
public:
    vector<int>* stack;
    vector<int>* min_stack;
    /** initialize your data structure here. */
    MinStack() {
        stack = new vector<int>();
        min_stack = new vector<int>();
        min_stack->push_back(INT_MAX);
    }
    
    void push(int x) {
        stack->push_back(x);
        int pre_min = min_stack->back();
        if (x < pre_min) 
            min_stack->push_back(x);     
        else
            min_stack->push_back(pre_min);
    }
    
    void pop() {
        stack->pop_back();
        min_stack->pop_back();
    }
    
    int top() {
        return stack->back();
    }
    
    int getMin() {
        return min_stack->back();
    }
};

File: python/155.py


MAX_VAL = 10000000000000

class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack = []
        self.min_stack = [MAX_VAL]


    def push(self, x: int) -> None:
        self.stack.append(x)
        if x < self.min_stack[-1]:
            self.min_stack.append(x)
        else:
            self.min_stack.append(self.min_stack[-1])

    def pop(self) -> None:
        self.stack.pop()
        self.min_stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min_stack[-1]